auth_mellon_util.c

```diff
@@ -1507,83 +1507,395 @@ am_get_service_url(request_rec *r, LassoProfile *profile, char *service_name)
+ * was successful. If an error occured the returned array of tokens
+ * will include all tokens parsed up until where the unrecognized
+ * input occurred. The input str is never modified.
  *
- * The first value must be a version declaration in the form ver="xxx"
- * (note the version string must be in double quotes).
+ * Parameters:
+ *  pool              memory allocation pool
+ *  str               input string to be parsed.
+ *  ignore_whitespace if True whitespace tokens are not returned
+ *  error             location where error string is returned
+ *                    if NULL no error occurred
+ * Returns:           array of Token objects
+ */
+
+#define TOKEN_INIT(t) memset(&t, 0, sizeof(t))
+
+#define PUSH_TOKEN(token_type, start, end)              \
+{                                                       \
+    const char *s = (start);                            \
+    const char *e = (end);                              \
+    apr_size_t offset = s - str;                        \
+                                                        \
+    if (token_type == TOKEN_DBL_QUOTE_STRING) {         \
+        /* do not include quotes in token value */      \
+        s++; e--;                                       \
+    }                                                   \
+                                                        \
+    token.type = token_type;                            \
+    token.len = e-s;                                    \
+    token.offset = offset;                              \
+    token.str = apr_pstrmemdup(pool, s, token.len);     \
+    APR_ARRAY_PUSH(tokens, Token) = token;              \
+                                                        \
+    TOKEN_INIT(token);                                  \
+}
+
+static apr_array_header_t *
+tokenize(apr_pool_t *pool, const char *str, bool ignore_whitespace,
+         char **error)
+{
+    apr_array_header_t *tokens = apr_array_make(pool, 10, sizeof(Token));
+    const char *p, *start;
+    Token token;
+
+    *error = NULL;
+    TOKEN_INIT(token);
+    p = start = str;
+    while(*p) {
+        if (apr_isspace(*p)) {  /* whitespace */
+            p++;
+            while(*p && apr_isspace(*p)) p++;
+            if (!ignore_whitespace) {
+                PUSH_TOKEN(TOKEN_WHITESPACE, start, p);
+            }
+            start = p;
+        }
+        else if (apr_isalpha(*p)) { /* identifier: must begin with
+                                       alpha then any alphanumeric or
+                                       underscore */
+            p++;
+            while(*p && (apr_isalnum(*p) || *p == '_')) p++;
+            PUSH_TOKEN(TOKEN_IDENTIFIER, start, p);
+            start = p;
+        }
+        else if (*p == '"') {   /* double quoted string,
+                                   quote escaped with backslash */
+            p++;
+            while(*p && (*p != '\"' || *(p-1) == '\\')) p++;
```

Comment:
I think you must have a slightly more complicated loop here, to deal with `"\\"`.